### Dataset explore

In [2]:
import os

image_dir = "../Dataset/JPEGImages"
xml_dir = "../Dataset/Annotations"

print("Images:", len(os.listdir(image_dir)))
print("XMLs:", len(os.listdir(xml_dir)))

Images: 5000
XMLs: 5000


### Open one XML

In [3]:
import xml.etree.ElementTree as ET

xml_file = os.path.join(xml_dir, os.listdir(xml_dir)[0])

tree = ET.parse(xml_file)
root = tree.getroot()

for elem in root.iter():
    print(elem.tag, elem.text)

annotation 
	
folder VOC2007
filename 000001.jpg
source 
		
database My Database
annotation VOC2007
image flickr
flickrid NULL
owner 
		
flickrid NULL
name xiaoxianyu
size 
		
width 718
height 500
depth 3
segmented 0
object 
		
name 0
pose Unspecified
truncated 0
difficult 0
bndbox 
			
xmin 223
ymin 90
xmax 286
ymax 131


### Master Image-Label Mapping Table

In [5]:
import os
import xml.etree.ElementTree as ET
import pandas as pd

# 1. Define where your files are located
# (Adjust these paths if your folders are located somewhere else)
xml_dir = "../Dataset/Annotations" 
output_dir = "../Outputs"

# Ensure your Output folder exists before saving files into it
os.makedirs(output_dir, exist_ok=True)

# 2. Create an empty list to store the data row-by-row
all_rows = []

# 3. Loop through every single file in your XML directory
print("Parsing XML files... Please wait.")
for file_name in os.listdir(xml_dir):
    # Only process files that end with .xml
    if file_name.endswith(".xml"):
        xml_path = os.path.join(xml_dir, file_name)
        
        # Parse the XML file tree structure
        tree = ET.parse(xml_path)
        root = tree.getroot()
        
        # Locate the <object> element
        obj = root.find("object")
        if obj is not None:
            # Extract the class label (0 or 1)
            label = obj.find("name").text
            
            # Locate the bounding box coordinates
            bndbox = obj.find("bndbox")
            xmin = int(bndbox.find("xmin").text)
            ymin = int(bndbox.find("ymin").text)
            xmax = int(bndbox.find("xmax").text)
            ymax = int(bndbox.find("ymax").text)
            
            # Determine the corresponding image file name
            image_name = file_name.replace(".xml", ".jpg")
            
            # Combine everything into a single row list
            row_data = [image_name, label, xmin, ymin, xmax, ymax]
            all_rows.append(row_data)

# 4. Convert the gathered list into a structured Pandas DataFrame
df = pd.DataFrame(
    all_rows, 
    columns=["image_name", "label", "xmin", "ymin", "xmax", "ymax"]
)

# 5. Save the DataFrame as a CSV spreadsheet
csv_save_path = os.path.join(output_dir, "image_label_mapping.csv")
df.to_csv(csv_save_path, index=False)

print(f"Success! Master mapping table created and saved to: {csv_save_path}")
print(f"Total entries processed: {len(df)}")

Parsing XML files... Please wait.
Success! Master mapping table created and saved to: ../Outputs\image_label_mapping.csv
Total entries processed: 5000


### Inspect Class Distribution

In [6]:
import pandas as pd

# Load the master file you just created
mapping_df = pd.read_csv("../Outputs/image_label_mapping.csv")

# Print out the counts of each unique label
print("--- Class Distribution Count ---")
print(mapping_df['label'].value_counts())

# Print a quick preview of the top 5 rows to ensure formatting is correct
print("\n--- Preview of Mapping Data Table ---")
print(mapping_df.head())

--- Class Distribution Count ---
label
1    3574
0    1426
Name: count, dtype: int64

--- Preview of Mapping Data Table ---
   image_name  label  xmin  ymin  xmax  ymax
0  000001.jpg      0   223    90   286   131
1  000002.jpg      0   488    90   539   130
2  000003.jpg      0   151   129   313   252
3  000004.jpg      0   331   115   441   185
4  000005.jpg      0   382   143   485   230


#### We found 
Class 1 (Malignant/Target): 3,574 images   
Class 0 (Benign/Control): 1,426 images   

### CV install  

In [7]:
# The '%' sign tells Jupyter to install this specifically into the environment your notebook is currently using.
%pip install opencv-python

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement opencv-python (from versions: none)

[notice] A new release of pip is available: 24.3.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip
ERROR: No matching distribution found for opencv-python


In [10]:
pip install --upgrade pip

   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/1.8 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/1.8 MB ? eta -:--:--
   ----------- ---------------------------- 0.5/1.8 MB 876.2 kB/s eta 0:00:02
   ----------- ---------------------------- 0.5/1.8 MB 876.2 kB/s eta 0:00:02
   ----------------- ---------------------- 0.8/1.8 MB 737.1 kB/s eta 0:00:02
   ----------------------- ---------------- 1.0/1.8 MB 731.3 kB/s eta 0:00:02
   ----------------------- ---------------- 1.0/1.8 MB 731.3 kB/s eta 0:00:02
   ----------------------- ---------------- 1.0/1.8 MB 731.3 kB/s eta 0:00:02
   ---------------------------- ----------- 1.3/1.8 MB 644.9 kB/s eta 0:00:01
   ---------------------------- --------